lspce: 1.0.0-unstable-2024-02-03 -> 1.1.0-unstable-2024-07-13

AndersonTorres opened the pull request:
## Description of changes

Fixing an Emacs package.

related: https://github.com/NixOS/nixpkgs/issues/278925

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jian-lin reviewed (commented):
Fixing `cargoHash` is enough to fix the build.

The old way of packaging makes more sense to me because `lspce` is an Emacs package which uses a dynamic module.

---

CI show another `cargoHash` [error][1].  Weird.

[1]: https://github.com/NixOS/nixpkgs/pull/326981/checks?check_run_id=27416440639

AndersonTorres commented:
> The old way of packaging makes more sense to me because lspce is an Emacs package which uses a dynamic module.

Hum, in a certain sense I agree. This package is not really useful outside Emacs.

However, the hardest work is made by Rust toolchain, while the Emacs part looks like a wrapper.
Indeed the modules can be built without any interference from Emacs ecosystem.
Further, technically the Emacs part depends on the Rust part.

Because of it I reworked the package this way: a Rust package with an Emacs companion package as passthru.

jian-lin commented:
> Indeed the modules can be built without any interference from Emacs ecosystem.
> Further, technically the Emacs part depends on the Rust part.

Every Emacs package using a dynamic module [depends on][1] `emacs-module.h` of Emacs to build.  `lspce` builds without explicit need of `emacs-module.h` because its core dependency, the `emacs` crate, [vendors][2] `emacs-module.h`.

[1]: https://www.gnu.org/software/emacs/manual/html_node/elisp/Module-Initialization.html
[2]: https://github.com/ubolonton/emacs-module-rs/commit/cad972872c61ad2b402a309d2c787b5195e0d503

jian-lin commented:
The version change in https://github.com/NixOS/nixpkgs/pull/316332 causes the change of `cargoHash`.  More info: https://github.com/NixOS/nixpkgs/issues/112763#issuecomment-777417991

jian-lin commented on pkgs/by-name/ls/lspce/package.nix:
```diff
@@ -0,0 +1,52 @@
+
+    passthru = {
+      elisp-package = callPackage ./elisp-package.nix {
+        inherit (emacsPackages) f markdown-mode melpaBuild yasnippet;
```
This way of introducing elisp dependencies is wrong because dependencies listed here is always from `emacsPackage` which equals to `pkgs.emacs.pkgs`.

For example, if `pkgs.emacs-nox` is used, these dependencies should be from `pkgs.emacs-nox.pkgs`.

AndersonTorres replied:
Oh, how could I forget the elisp packages not being retrieved from Hydra I reported since ~~Heian era~~ https://github.com/NixOS/nixpkgs/issues/324912 ?...

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/lspce/module.nix:
```diff
@@ -0,0 +1,47 @@
+let
+  self = rustPlatform.buildRustPackage {
+    pname = "lspce-module";
+    version = "1.1.0-unstable-2024-07-13";
```
Can we still use `unstableGitUpdater` to auto-update?

AndersonTorres replied:
I don't think so. The updaters are not so smart to look outside their files, I believe.

But maybe we can modify the updaters to do something like "hey, don't look at me, look to this file, the hashes are there".

AndersonTorres commented:
Apologies for the delay, I was dealing with some [orphans](https://github.com/NixOS/nixpkgs/issues/327779) and [corpses](https://github.com/NixOS/nixpkgs/issues/290642).

I took a middle approach:

- The lspce-module has its own file;
- The elisp code calls this module.

jian-lin reviewed (changes requested):
`lspce.el` fails to compile.

My fix (with some minor improvements such as adding an update script) is available here: https://github.com/atorres1985-contrib/nixpkgs/compare/6837f904c5fad033ff35074903953c0de2a727bf...linj-fork:nixpkgs:tmp-pr/326981

If you are busy, I am willing to create a new PR because it seems that I cannot push to this specific PR for some reason.

jian-lin commented:
Ideally, the bump and the refactor should be two separate PRs.

AndersonTorres committed:
emacsPackages.lspce: rework

- detach Rust module to module.nix
- trivialBuild -> melpaBuild
- updateScript
- adopt (by AndersonTorres)

Co-authored-by: Lin Jian <me@linj.tech>

AndersonTorres committed:
emacsPackages.lspce: 1.0.0-unstable-2024-02-03 -> 1.1.0-unstable-2024-07-13

jian-lin reviewed (approved):
LGTM
